### PR TITLE
Couple two phase flow with heat transfer

### PR DESCRIPTION
**Pull request by mschreter**
This PR includes:
- add `two_phase_flow_with_heat_transfer` problem; If this problem type is chosen, a `HeatTransferOperation` is instantiated in the `TwoPhaseFlowProblem`.
- option enabled for solving a heat problem in case of a melt pool problem
- `compute_liquid_and_solid_phases`: liquid and solid regions are now distinguished based on the level set value and the temperature value.

Due to the numerous extensions available in the `TwoPhaseFlowProblem` it is currently a "mess". We should definitely discuss to split it up in separate operations, i.e. two-phase-flow, two-phase-flow with phase change, to maintain a clear coding structure.

### CODE REVIEW

**Comment by mschreter** [comment]
@peterrum: Would you mind, giving this PR a quick look? Sorry for the quick and dirty implementation. TwoPhaseFlowProblem/MeltPoolOperation are shouting for refactoring!!!

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 1]
```diff
@@ -123,12 +123,14 @@ namespace MeltPoolDG
...
-                                                                data_in.laser.gusarov,
-                                                                temp_dof_idx);
+        if (mp_data.temperature_formulation == "solve_heat_equation" &&
+            data_in.laser.heat_source_model == "Gusarov")
+          {
+            laser_heat_source_operation =
+              std::make_shared<Heat::LaserHeatSourceGusarov<dim>>(*scratch_data,
+                                                                  data_in.laser.gusarov,
+                                                                  temp_dof_idx);
+          }
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:133]: else? Isn't an assert missing?

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 2]
```diff
@@ -149,6 +151,10 @@ namespace MeltPoolDG
             density_liquid,
             mp_data,
             -1.0 /* level set value for gas @todo */);
+        else if (mp_data.temperature_formulation == "solve_heat_equation")
+          {
+            heat_operation->get_temperature() = mp_data.ambient_temperature;
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:156]: The ambient temperature is the initial condition?

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 3]
```diff
@@ -305,11 +304,30 @@ namespace MeltPoolDG
       }
 
     private:
+      void
+      make_constraints_in_spatially_fixed_solid_domain()
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:308]: This function has been made private. Nothing else has changed!?

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 4]
```diff
@@ -458,30 +480,29 @@ namespace MeltPoolDG
...
-        if (point[dim - 1] >= laser_operation->get_laser_position()[dim - 1])
-          return false;
+        if (phi_liquid <= 0.5)
+          return false; // point is gas phase
+        else if (phi_liquid > 0.5 && temperature >= mp_data.liquid.melting_point)
+          return false; // point is melted
         else
-          return (temperature >= mp_data.liquid.melting_point) ? false : true;
+          return true;
       }
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:493]: Could you explain the logic of this function and the function below. How is phi related to solid?

File: include/meltpooldg/heat/laser_heat_source_gusarov.hpp [hunk 5]
```diff
@@ -148,7 +148,7 @@ namespace MeltPoolDG::Heat
      * volumetric heat source; The z-axis (= axis of the laser beam) is assumed to correspond to
      * negative dim-1 coordinate.
      */
-    void
+    double
```
Comment (peterrum) [inline include/meltpooldg/heat/laser_heat_source_gusarov.hpp:151]: :)

**Review by peterrum — COMMENTED** [review]
Mainly some questions.

File: include/meltpooldg/heat/laser_heat_source_gusarov.hpp (hunk 5 above)
Comment (mschreter) [reply include/meltpooldg/heat/laser_heat_source_gusarov.hpp:151]: Never used this function before ^^.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp (hunk 2 above)
Comment (mschreter) [reply include/meltpooldg/melt_pool/melt_pool_operation.hpp:156]: For the moment, yes.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp (hunk 4 above)
Comment (mschreter) [reply include/meltpooldg/melt_pool/melt_pool_operation.hpp:493]: done.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp (hunk 3 above)
Comment (mschreter) [reply include/meltpooldg/melt_pool/melt_pool_operation.hpp:308]: No, nothing else.

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp (hunk 1 above)
Comment (mschreter) [reply include/meltpooldg/melt_pool/melt_pool_operation.hpp:133]: I've added an assert.

**Comment by mschreter** [comment]
Thank you Peter, I've adopted your suggestions. All tests pass locally. If you are happy with that feel free to merge :-).

File: include/meltpooldg/melt_pool/melt_pool_operation.hpp [hunk 6]
```diff
@@ -123,366 +123,386 @@ namespace MeltPoolDG
...
-          return (temperature >= mp_data.liquid.melting_point) ? true : false;
-      }
-    };
-  } // namespace MeltPool
-} // namespace MeltPoolDG
+        /**
+         *  This function determines for a given pair of level set and temperature values, whether
+         * it characterizes a solid state. For this purpose, it is assumed that phi=1 in the liquid
+         * AND the solid domain and phi=0 in the gaseous domain. Thus, if phi=1 and the temperature
+         * is SMALLER than the fusion point, a solid phase is met.
```
Comment (peterrum) [inline include/meltpooldg/melt_pool/melt_pool_operation.hpp:480]: "Solid phase is Met." :clinking_glasses:

**Review by peterrum — APPROVED** [review]

**Pull request merged**